lib/utils/patterns.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+var RE2;
+var minimatch = require('minimatch');
+try {
+  RE2 = require('re2');
+} catch (er) {
+  RE2 = null;
+}
+
+var patterns = {
+  /**
+   * Evaluates whether a configured redirect/rewrite/custom header should
+   * be applied to a request against a specific path. All three features
+   * are configured with a hash that contains either a Node-like glob path
+   * specification as its `source` or `glob` field, or a RE2 regular
+   * expression as its `regex` field.
+   *
+   * Since Javascript lacks a native library for RE2, Superstatic uses the C
+   * bindings as an optional dependency, and falls over to PCRE if the import
+   * is unavailable. Under most circumstances not involving named capturing
+   * groups, the two libraries should have identical behavior.
+   *
+   * No special consideration is taken if the configuration hash contains both
+   * a glob and a regex. normalizeConfig() will error in that case.
+   *
+   * @param {string} path The URL path from the request.
+   * @param {Object} config A dictionary from a sanitized JSON configuration.
+   * @return {boolean} Whether the config should be applied to the request.
+   */
+  configMatcher: function(path, config) {
+    var glob = config.glob || config.source;
+    var regex = config.regex;
+    if (glob && minimatch(path, glob)) {
+      return true;
+    }
+    if (regex && RE2) {
+      return (path.match(new RE2(regex)) !== null);
+    }
+    if (regex && !RE2) {
+      return (path.match(new RegExp(regex)) !== null);
+    }
+    return false;
+  },
+
+  /**
+   * Creates either an RE2 or a Javascript RegExp from a provided string
+   * pattern, depending on whether or not the RE2 library is available as an
+   * import.
+   *
+   * @param {string} pattern A regular expression pattern to test against.
+   * @return {regexp} A regular expression object, created by either base
+   *                  RegExp or RE2, which matches the RegExp prototype
+   */
+  createRaw: function(pattern) {
+    if (RE2) {
```

Comment:
Again, let's re-use logic instead of sprinkling the same code in many places.

```js
    if (re2Available()) {
```

Reply:
I'm kind of resisting this because I like inlining exported functions which means they can't directly call each other, but also these if blocks can be compressed a lot by using ternary operators. What do you think? Otherwise, it's pretty simple to copy-paste them all into their own non-anonymous functions and then just export the hash at the end of the file.